### PR TITLE
typing: add types to beets.util and beets.test.helper

### LINKED ISSUE #6808: Make path utilities generic over string path types
Parent issue: #1409

## Problem

Many beets path utilities are typed or implemented around byte paths and sometimes coerce inputs to `bytes`. This prevents callers from migrating incrementally to string and `Path` representations because a helper may reject the new input type or unexpectedly return bytes.

## Expected behavior

- Make path-producing helpers generic over `str` and `bytes`, preserving the caller's concrete path type in returned or yielded values.
- Broaden filesystem-action helpers to accept `PathLike` when their return type does not depend on the input path type.
- Apply this to the utilities including `mkdirall`, `samefile`, `copy`, `move`, `link`, `hardlink`, `reflink`, `case_sensitive`, `unique_path`, `sorted_walk`, `hidden.is_hidden`, and `albums_in_dir`.
- Keep ignore patterns, regular expressions, path components, and recursive calls type-consistent so bytes and strings are never mixed at runtime.
- Retain explicit byte conversion only at boundaries whose contracts require bytes.
- Preserve existing filesystem behavior.

## Acceptance criteria

- Each affected utility accepts the intended string, byte, or general path-like inputs without type errors.
- Helpers whose outputs are derived from their inputs preserve `str` or `bytes` consistently.

This work must land before migrating the broader test suite to `pathlib.Path`.

### PR DESCRIPTION
**Pull request by snejus**
I've started working on migrating the codebase to `pathlib.Path` and this is the very first step: fully type `beets.util` and `beets.test.helper`. Next, I will me migrating tests to use `pathlib.Path`.

Fixes: #6808

- Extracts shared path/temp-directory behavior into `PathsMixin`, and splits import-specific test setup into `ImporterMixin`. This makes `beets.test.helper` more modular and gives test helpers a clearer inheritance structure.

- Updates `TestHelper` and import helpers to use those smaller mixins, while keeping existing test behavior intact. The main architectural effect is better separation between core test setup, filesystem helpers, and importer-specific utilities.

- Tightens typing across `beets.test.helper` and `beets.util`, especially around path handling. `beets.util` now accepts a broader set of path types through `PathLike`/`AnyPath`, which reduces friction between `bytes`, `str`, and `Path` usage.

- Aligns helper internals with the typed interfaces by normalizing path handling in methods like `touch`, using typed `template(...)` entries for importer path formats, and fixing a few return-type/inheritance edge cases.

- High-level impact: this is mainly a typing and test-infrastructure cleanup. It improves maintainability and type-checker accuracy without changing production architecture or intended runtime behavior.

### CODE REVIEW

**Review by semohr — APPROVED** [review]
Looks good from my perspective.

Consider adding the files to the optin stricter mypy config.

**Comment by snejus** [comment]
> Consider adding the files to the optin stricter mypy config.

I tried adding them:

```diff
diff --git a/setup.cfg b/setup.cfg
index 01971c01d..259e8a5c0 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -55,6 +55,12 @@ check_untyped_defs = true
 disallow_untyped_decorators = true
 check_untyped_defs = true
 
+[[mypy-beets.utils]]
+strict = true
+
+[[mypy-beets.test.helper]]
+strict = true
+
 [[mypy-beetsplug.musicbrainz]]
 strict = true
```

And ran `mypy` on `master` but it didn't complain about any issues :/

**Comment by semohr** [comment]
> Consider adding the files to the optin stricter mypy config.

Maybe using `disallow_untyped_defs` would work? Am also not sure the `config` is picked up correctly :thinking: 

My idea here was just to have future changes require stricter rules for these files as they are typed properly right now.

**Comment by snejus** [comment]
> Maybe using `disallow_untyped_defs` would work?

Tried this but `mypy` is still silent :/

```diff
diff --git a/beets/util/__init__.py b/beets/util/__init__.py
index 52dcd1833..a9f7e0d14 100644
--- a/beets/util/__init__.py
+++ b/beets/util/__init__.py
@@ -1011,7 +1011,7 @@ def case_sensitive
         return not os.path.samefile(lower_sys, upper_sys)
 
 
-def asciify_path(path: str) -> str:
+def asciify_path(path) -> str:
     """Decodes all unicode characters in a path into ASCII equivalents.
 
     Substitutions are provided by the unidecode module. Path separators in the
diff --git a/setup.cfg b/setup.cfg
index 01971c01d..2f17cf07b 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -55,6 +55,12 @@ check_untyped_defs = true
 disallow_untyped_decorators = true
 check_untyped_defs = true
 
+[[mypy-beets.util]]
+allow_untyped_defs = false
+
+[[mypy-beets.test.helper]]
+allow_untyped_defs = false
+
 [[mypy-beetsplug.musicbrainz]]
 strict = true
```

**Comment by snejus** [comment]
On the other hand I myself use `ruff check --select=ANN` (`flake8-annotations`) for this.

File: beets/test/helper.py [hunk 1]
```diff
@@ -668,31 +697,38 @@ class ImportSessionFixture(ImportSession):
...
-        if isinstance(choice, int):
-            return task.candidates[choice - 1]
+        if task.candidates:
+            if choice == importer.Action.APPLY:
+                return task.candidates[0]  # type: ignore[return-value]
+            if isinstance(choice, int):
+                return task.candidates[choice - 1]  # type: ignore[return-value]
+
+        assert not isinstance(choice, int), f"Invalid choice: {choice}"
         return choice
```
[the remark this answers is not in the document]
Comment (snejus) [reply beets/test/helper.py:729]: This behaviour existed before.

**Comment by snejus** [comment]
I'll merge it for now until we have a resolution for `mypy`. There are a couple of fully-typed modules now that the setting should apply to!
